src/Common/src/Microsoft/Win32/SafeHandles/NtlmSecuritySafeHandles.cs

```diff
@@ -0,0 +1,271 @@
+            : base(IntPtr.Zero, true)
+        {
+            string keyMagic = string.Format(s_keyMagic, isClient ? s_client : s_server,
+                    isClient ? s_server : s_client, isSealingKey ? s_sealing : s_signing);
+
+            using (IncrementalHash incremental = IncrementalHash.CreateHash(HashAlgorithmName.MD5))
+            {
+                 incremental.AppendData(key);
+                 incremental.AppendData(Encoding.UTF8.GetBytes(keyMagic));
+                 _digest = incremental.GetHashAndReset();
+            }
+
+            _isSealingKey = isSealingKey;
+            if (_isSealingKey)
+            {
+                _cipherContext = Interop.Crypto.EvpCipherCreate(Interop.Crypto.EvpRc4(), _digest, null, 1);
+            }
+        }
+
+        public override bool IsInvalid
+        {
+            get { return true; }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && (null != _cipherContext))
+            {
+                _cipherContext.Dispose();
+                _cipherContext = null;
+            }
+            base.Dispose(disposing);
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            return true;
+        }
+
+        public byte[] Sign(SafeNtlmKeyHandle sealingKey, byte[] buffer, int offset, int count)
+        {
+            Debug.Assert(!_isSealingKey, "Cannot sign with sealing key");
+            Debug.Assert(offset >= 0 && offset < buffer.Length, "Cannot sign with invalid offset " + offset);
+            Debug.Assert((count + offset) <= buffer.Length, "Cannot sign with invalid count " + count);
+
+            // reference for signing a message: https://msdn.microsoft.com/en-us/library/cc236702.aspx
+            const uint Version = 0x00000001;
+            const int ChecksumOffset = 4;
+            const int SequenceNumberOffset = 12;
+            const int HMacDigestLength = 8;
+
+            byte[] output = new byte[Interop.NetSecurity.MD5DigestLength];
+            Array.Clear(output, 0, output.Length);
+            MarshalUint(output, 0, Version); // version
+            MarshalUint(output, SequenceNumberOffset, _sequenceNumber);
+            byte[] hash;
+
+            using (var incremental = IncrementalHash.CreateHMAC(HashAlgorithmName.MD5, _digest))
+            {
+                incremental.AppendData(output, SequenceNumberOffset, ChecksumOffset);
+                incremental.AppendData(buffer, offset, count);
+                hash = incremental.GetHashAndReset();
+                _sequenceNumber++;
+            }
+
+            if (sealingKey == null)
+            {
+                Array.Copy(hash, 0, output, ChecksumOffset, HMacDigestLength);
+            }
+            else
+            {
+                byte[] cipher = sealingKey.SealOrUnseal(true, hash, 0, HMacDigestLength);
+                Array.Copy(cipher, 0, output, ChecksumOffset, cipher.Length);
```

Comment:
Rather than allocating a new array simply to then copy it to a target array, should SealOrUnseal just write directly into the target array?